app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/fragments/PopupAttributesListFragment.kt

```diff
@@ -43,11 +50,69 @@ class PopupAttributesListFragment : Fragment() {
             false
         )
 
-        val viewModel: IdentifyResultViewModel by activityViewModels()
+        val identifyResultViewModel: IdentifyResultViewModel by activityViewModels()
 
-        binding.identifyResultViewModel = viewModel
+        binding.identifyResultViewModel = identifyResultViewModel
+        binding.popupViewModel = popupViewModel
         binding.lifecycleOwner = this
 
+        popupViewModel.isPopupInEditMode.observe(viewLifecycleOwner, Observer {
+            popupView.setEditMode(it)
+        })
+
+        popupViewModel.toggleSavingPopupProgressBarEvent.observeEvent(viewLifecycleOwner) { isShowProgressBar ->
+            if (isShowProgressBar) {
+                progressBarLayout.visibility = View.VISIBLE
+            } else {
+                progressBarLayout.visibility = View.GONE
+            }
+        }
+
+        popupViewModel.showSavePopupErrorEvent.observeEvent(viewLifecycleOwner) { errorMessage ->
+            showAlertDialog(errorMessage)
+        }
+
+        popupViewModel.confirmCancelPopupEditingEvent.observeEvent(viewLifecycleOwner) {
+            showConfirmCancelEditingDialog()
+        }
+
         return binding.root
     }
+
+    /**
+     * Shows dialog to confirm cancelling edit mode on popup view.
+     */
+    private fun showConfirmCancelEditingDialog() {
+        val dialogBuilder = AlertDialog.Builder(requireContext())
+        dialogBuilder.setMessage("Discard changes?")
+            .setCancelable(false)
+            // positive button text and action
+            .setPositiveButton(getString(R.string.ok)) { dialog, id ->
+                popupViewModel.cancelEditing()
+            }
+            // negative button text and action
+            .setNegativeButton(getString(R.string.cancel)) { dialog, id -> dialog.cancel()
+            }
+        val alert = dialogBuilder.create()
+        // show alert dialog
+        alert.show()
+    }
+
+    /**
+     * Shows error message to the use in a dialog.
+     */
+    private fun showAlertDialog(message: String) {
```

Comment:
I wonder if the two "show dialog" methods could be combined into one. Ideally put them in a helper class so it can be re-used in other places.

Reply:
The two show dialogs cannot be combined into one as one has just the cancel button(it dismisses the dialog when user presses it) and the second one has both OK and Cancel. 

In the 1st case when the user presses OK we just dismiss the dialog. 
In the 2nd case when the user presses OK we need to perform some follow up operation when user presses OK. 

The solution to get some value back when user presses the button is not very straight forward,
https://stackoverflow.com/questions/10763986/android-confirmation-dialog-returning-true-or-false